### PR TITLE
Added Merge Sort [Javascript]

### LINKED ISSUE #375: Merge Sort [Javascript]
#### This issue is about [Merge Sort](https://www.khanacademy.org/computing/computer-science/algorithms/merge-sort/a/overview-of-merge-sort).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by rashinanwani**
Fixes #375

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

File: merge_sort/merge_sort.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+function mergeSort (arr) {
+	/*
+	Dividing the array into two parts, sorts both the part of array and then merge it
+	Time complexity : O(size(log(size)))
+	Space complexity : O(size)
+	*/
+	if (arr.length < 2) {
+		return arr;
+	}
+	let middle = arr.length / 2;
```
Comment (Anshuman-Verma) [inline merge_sort/merge_sort.js]: middle value evaluates to a decimal value for arrays of odd length which does not seem right, use `//` i.e. strict division to return an integer off it.
Rest all looks good.

**Comment by Anshuman-Verma** [comment]
Check comments please.

**Comment by Anshuman-Verma** [comment]
Also use camelCase for naming js files as stated in #333.
In this case `mergeSort.js`  :smiley:

File: merge_sort/mergeSort.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/* 
+*	Implementation of merge sort algorithm
+*	Time complexity: O(size(log(size)))
```
Comment (Anshuman-Verma) [inline merge_sort/mergeSort.js]: Why turned it to `size(log(size))` ? 
`log(size)` was sufficient i think.

File: merge_sort/mergeSort.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+*	Implementation of merge sort algorithm
+*	Time complexity: O(size(log(size)))
+*	Space complexity: O(size)
+*/
+function merge (left, right) {
+	/*
+	*	:Merging two sorted array
+	*	:param arr: array.
```
Comment (aviaryan) [inline merge_sort/mergeSort.js]: There are no params like `arr` and `size`.

File: merge_sort/mergeSort.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/*
+*	Implementation of merge sort algorithm
+*	Time complexity: O(size(log(size)))
```
Comment (aviaryan) [inline merge_sort/mergeSort.js]: Making it `size * log(size)` will make it more clear.

File: ; [hunk 5]
```diff
@@ -0,0 +1,25 @@
+pick ec42def Added Merge Sort [Javascript] Fixes #375
```
Comment (Anshuman-Verma) [inline ;]: Extra file is not needed

**Comment by Anshuman-Verma** [comment]
@rashinanwani it's been days since this was last active, kindly make changes.

File: merge_sort/mergeSort.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+			arr[k] = right[j];
+			j++;
+		}
+	}
+	return arr;
+}
+
+function mergeSort (arr) {
+	/*
+	*	:param arr: array to be sorted
```
Comment (aviaryan) [inline merge_sort/mergeSort.js]: Docstrings, like this, usually go above the function declaration.

Comment (aviaryan) [inline merge_sort/mergeSort.js]: @rashinanwani This issue isn't resolved yet. cc @Anshuman-Verma

**Comment by Anshuman-Verma** [comment]
@aviaryan can you please explain the expected comment structure ?

**Comment by aviaryan** [comment]
@Anshuman-Verma @rashinanwani  I was just asking to move the doc-string above the function. (See [comment](https://github.com/iiitv/algos/pull/399#discussion_r127597102))

```js
/*
 * function comments
 */
function name(){
    alert('foo');
}
```

**Comment by Anshuman-Verma** [comment]
@aviaryan thanks for the follow up, missed that review comment.
@rashinanwani make the required changes please.

**Pull request merged**